pkg/catalog/catalog.go

```diff
@@ -1720,6 +1727,145 @@ func (c *Catalog) PrepareExpiredCommits(ctx context.Context, repositoryID string
+	var buf bytes.Buffer
+	encoder := base64.NewEncoder(base64.StdEncoding, &buf)
+	err := json.NewEncoder(encoder).Encode(m)
+	if err != nil {
+		return "", err
+	}
+	result := buf.String()
+	return result, nil
+}
+
+type UncommittedParquetObject struct {
+	PhysicalAddress string `parquet:"name=physical_address, type=BYTE_ARRAY, convertedtype=UTF8, encoding=PLAIN_DICTIONARY"`
+	CreationDate    int64  `parquet:"name=creation_date, type=INT64, convertedtype=INT_64"`
+}
+
+func (c *Catalog) writeUncommittedLocal(ctx context.Context, repository *graveler.RepositoryRecord, fd *os.File, mark *GCUncommittedMark) (*GCUncommittedMark, error) {
+	const parallelNum = 1                                                                        // Number of goroutines to handle marshaling of data
+	pw, err := writer.NewParquetWriterFromWriter(fd, new(UncommittedParquetObject), parallelNum) // TODO: Play with np count
+	if err != nil {
+		return nil, err
+	}
+	// Compression type
+	pw.CompressionType = parquet.CompressionCodec_GZIP
+
+	const GCMaxUncommittedFileSize = 20 * 1024 * 1024 // 20 MB
+	// Calculation of size deviation by periodicCheckSize value
+	// "data" prefix = 4 bytes
+	// partition id = 20 bytes (xid)
+	// object name = 20 bytes (xid)
+	// timestamp (int64) = 8 bytes
+	//
+	// Total per entry ~52 bytes
+	// Deviation with periodicCheckSize = 100000 will be around 5 MB
+	const periodicCheckSize = 100000
+	fileSizeExceeded := func(fd *os.File) (bool, error) {
+		stat, err := fd.Stat()
+		if err != nil {
+			return false, err
+		}
+		return stat.Size() > GCMaxUncommittedFileSize, nil
+	}
+
+	count := 0
+	itr, err := NewUncommittedIterator(ctx, c.Store, repository)
+	if err != nil {
+		return nil, err
+	}
+	defer itr.Close()
+
+	if mark != nil {
+		itr.SeekGE(mark.BranchID, mark.Path)
+	}
+
+	for itr.Next() {
+		entry := itr.Value()
+		count += 1
+		if count%periodicCheckSize == 0 {
+			err := pw.Flush(true)
+			if err != nil {
+				return nil, err
+			}
+			done, err := fileSizeExceeded(fd)
+			if err != nil {
+				return nil, err
+			}
+			if done {
+				if itr.Err() != nil {
+					return nil, itr.Err()
+				}
+				err = pw.WriteStop()
+				if err != nil {
+					return nil, err
+				}
+				return &GCUncommittedMark{
+					BranchID: entry.branchID,
+					Path:     entry.Path,
+				}, nil
+			}
+		}
+		if err = pw.Write(UncommittedParquetObject{
+			PhysicalAddress: entry.Address,
+			CreationDate:    entry.LastModified.AsTime().Unix(),
+		}); err != nil {
+			return nil, err
+		}
+	}
+	if itr.Err() != nil {
+		return nil, itr.Err()
+	}
+	err = pw.WriteStop()
+	if err != nil {
+		return nil, err
+	}
+
+	// Finished reading all staging area - no continuation token
+	return nil, nil
+}
+
+// TODO (niro): Add check in controller that runID is provided if mark != nil
+func (c *Catalog) PrepareGCUncommitted(ctx context.Context, repositoryID string, runID *string, mark *GCUncommittedMark) (*graveler.GarbageCollectionRunMetadata, *GCUncommittedMark, error) {
```

Comment:
Suggestions:
1. Return PrepareGCUncommittedInfo or a single struct instead of multiple return variables.
2. RunID as string when empty is the same as passing nil